### PR TITLE
virt-handler canary upgrade

### PR DESCRIPTION
**Pull request by acardace**
**What this PR does / why we need it**:

This PR introduces a new upgrade mechanism to first safely upgrade one virt-handler (one node) and then proceed with rest in batches, making the upgrade procedure faster (but still safe) on large clusters.

This upgrade mechanism ensures that virt-handler is safe to upgrade
by launching one (updated) canary virt-handler and making sure that the pod is
fully functional, after that we set `MaxUnavailable` to 10% and start
the real rollout that will execute on batches of 10% of nodes at a time
thus making the daemonset rollout on large cluster (> 20 nodes) 10 times faster.

If the canary-pod fails the virt-handler daemonset is rolledback to a
previous working revision and one warning event is created.

The canary-pod upgrade procedure in steps:
- update virt-handler with maxUnavailable=1
- patch daemonSet with new version
- wait for a new virt-handler to be ready
- set maxUnavailable=10%
- start the rollout of the new virt-handler again
- wait for all nodes to complete the rollout
- set maxUnavailable back to 1

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
implement virt-handler canary upgrade and rollback for faster and safer rollouts
```

### CODE REVIEW

File: pkg/util/rollback/rollback.go
Comment (davidvossel) [inline pkg/util/rollback/rollback.go:96]: in the event of a canary failure, I don't think the virt-operator should take any automated action to roll anything back. The admin has "declared" the version of KubeVirt they want. The virt-operator should block (and report) when it can't converge on the desired state. It's then the admin's choice to determine how to proceed.

With the canary pod we're simply attempting to reduce the blast radius, not make any decisions about how to correct the issue.

File: pkg/virt-operator/util/canary.go
Comment (davidvossel) [inline pkg/virt-operator/util/canary.go:44]: I don't think you need this annotation, You're just looking for a pod in a "ready" state that's using the updated image.

File: pkg/virt-operator/resource/apply/apps.go
Comment (davidvossel) [inline pkg/virt-operator/resource/apply/apps.go:158]: it seems like this should be as simple as

```
if updatedAndReadyPods == 0 {
    set maxUnavailable = 1
} else if updatedAndReadyPods > 0 && updatedAndReadyPods < totalExpected {
   Set maxUnavailable = 10%
} else if updatedAndReadyPods == totalExpected {
  Set maxUnavailable = 1%
}
```

The number of `updatedAndReadyPods` can be calculated by looking at virt-handler pods in the pod cache.

**Review by davidvossel — COMMENTED** [review]
The steps you have in the description look accurate however I think the execution can be simplified. For example, I don't think we need rollback, and we shouldn't need to an annotation to signify a "canary" pod either.

File: pkg/util/rollback/rollback.go
Comment (acardace) [reply pkg/util/rollback/rollback.go:96]: So no rollback at all? I guess it makes sense to not take any automated action against what the user declared but it kinda sounded nice to have as it  would restore the node to a functioning state.

Comment (acardace) [reply pkg/util/rollback/rollback.go:96]: If the rollback is not needed at all then everything becomes simpler, I probably misunderstood as I took it as part of the specs to solve the slow upgrade issue.

Comment (davidvossel) [reply pkg/util/rollback/rollback.go:96]: yup, automated rollback isn't necessary. If the upgrade encounters an issue, the expectation is that the issue is reported (via condition, events, etc...) and that intervention is required to decide how to correct the issue. Someone could create their own automation to rollback kubevirt if update is blocked or takes too long, but our virt-operator will only try to continually converge on the declared intent within the KubeVirt CR.

Comment (acardace) [reply pkg/util/rollback/rollback.go:96]: Right, got it, I'll modify and re-push, thank you for taking a look!

**Comment by acardace** [comment]
@davidvossel I've simplified everything by removing the rollback, please take a look.

**Comment by davidvossel** [comment]
it looks like unit tests are failing

File: pkg/virt-operator/resource/apply/apps.go
Comment (davidvossel) [inline pkg/virt-operator/resource/apply/apps.go:220]: I'm having trouble with this switch statement. Particularly around the cases where no action is taking place.

If we're going to use a switch statement, can the cases map directly to the steps? Something like... 

```
case updatedAndReadyPods == 0:
   // set max unavailable to 1 if it isn't already
   // else no-op and wait 
  return false
case updatedAndReadyPods > 0 && updatedAndReadyPods < desiredReadyPods 
  // set max unavailable to 10% if it isn't already
  // else no-op and wait
  return false
case updatedAndReadyPods > 0 && updatedAndReadyPods == desiredReadyPods
  // set max unavailable to 1 if it isn't already
  // set generation
  // daemonset is fully online, return so return true
  return true
```

Comment (acardace) [reply pkg/virt-operator/resource/apply/apps.go]: Updated, does this look better?

**Comment by acardace** [comment]
> it looks like unit tests are failing

Thanks, they should be fixed now.

File: pkg/virt-operator/resource/apply/apps.go
Comment (davidvossel) [inline pkg/virt-operator/resource/apply/apps.go:196]: for consistency, should this be `setMaxUnavailable(newDS, daemonSetDefaultMaxUnavailable)` instead of 1?

**Review by davidvossel — COMMENTED** [review]
/approve

excellent work! thanks for sticking with this. 

I made one minor comment about consistency of how the initial maxUnavailable value is set, but that's about it from a feature standpoint.

From a functional test standpoint, it would be nice if we could at least observe the daemonset going from `1` to `10%` back to `1`.  What would you think about making a small change to the virt-handler using the Customize Patch logic, and having an informer in the test catch those various stages as the daemonset rolls out?

**Comment by davidvossel** [comment]
/hold

I gave this an approve because I think the code is solid. I'd like to think through the functional tests a bit more before this is merged though.

File: pkg/virt-operator/resource/apply/apps.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/apps.go:196]: Will change it in the next push, thanks for spotting this!

**Comment by acardace** [comment]
@davidvossel added e2e test

File: tests/canary_upgrade_test.go
Comment (davidvossel) [inline tests/canary_upgrade_test.go:185]: the expectedEvents slice is being accessed by two goroutines at the same time without being protected by a mutex. This could cause issues since it's not technically thread safe.

Can you add a mutex around the `expectedEvents = processDsEvent(ds, expectedEvents)` and `len(expectedEvents) == 0`

Comment (acardace) [reply tests/canary_upgrade_test.go]: Thanks for spotting this, it is indeed something to protect with a mutex, already pushed the change.

**Comment by stu-gott** [comment]
@davidvossel your hold from 2 weeks ago is still present. Have your concerns been met?

**Review by davidvossel — APPROVED** [review]
/lgtm
/hold cancel

**Comment by acardace** [comment]
@davidvossel I just rebased, can you re-lgtm?

**Review by xpivarc — APPROVED** [review]
Granting LGTM based on David's review and a fact that the last push contains only correction of `serial`  -> `Serial`

File: tests/canary_upgrade_test.go
Comment (xpivarc) [inline tests/canary_upgrade_test.go:72]: Don't we need to wait for handlers to be ready again?

**Review by xpivarc — COMMENTED** [review]
/hold
@acardace I think there is a probability that this would flake. 
We fixed the `serial` "bug" but we didn't address that this test can affect the following test.

Last question: Should this be part of the operator lane?

File: tests/canary_upgrade_test.go
Comment (acardace) [reply tests/canary_upgrade_test.go:72]: You're right we probably should.

Comment (acardace) [reply tests/canary_upgrade_test.go:72]: Fixed it!

File: pkg/virt-operator/resource/apply/apps.go
Comment (xpivarc) [inline pkg/virt-operator/resource/apply/apps.go:250]: Wouldn't `daemonSetDefaultMaxUnavailable` be more precise?

File: pkg/virt-operator/resource/apply/apps.go
Comment (xpivarc) [inline pkg/virt-operator/resource/apply/apps.go:203]: I would break the assumption here that there will be only one matching pod.

**Review by xpivarc — APPROVED** [review]
Thanks for your work! I have only a few suggestions that I leave up to you.

File: pkg/virt-operator/resource/apply/apps.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/apps.go:250]: Fixed.

File: pkg/virt-operator/resource/apply/apps.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/apps.go:203]: Fixed.

File: pkg/virt-operator/resource/apply/apps.go
Comment (acardace) [reply pkg/virt-operator/resource/apply/apps.go:181]: I'd keep it like this, IMO the code gets a tiny bit less readable with named variables in this case.

File: tests/canary_upgrade_test.go
Comment (acardace) [reply tests/canary_upgrade_test.go:72]: @xpivarc can you unhold now that this is fixed?

**Pull request merged**
